controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -118,9 +118,31 @@ else if (retryBecauseNewFailure(application, report.jobType()))
     /** Returns whether all production zones listed in deployment spec last were successful on the currently deploying change. */
     private boolean deploymentComplete(LockedApplication application) {
         if ( ! application.deploying().isPresent()) return true;
-        return order.jobsFrom(application.deploymentSpec()).stream()
-                .filter(JobType::isProduction)
-                .allMatch(jobType -> application.deploymentJobs().isSuccessful(application.deploying().get(), jobType));
+        Change change = application.deploying().get();
+
+        for (JobType job : order.jobsFrom(application.deploymentSpec())) {
+            if ( ! job.isProduction()) continue;
+
+            Optional<ZoneId> zone = job.zone(this.controller.system());
+            if ( ! zone.isPresent()) continue;
+
+            Deployment deployment = application.deployments().get(zone.get());
+            if (deployment == null) return false;
+
+            // Check actual job outcome (the deployment)
+            if (change instanceof VersionChange) {
+                if (((VersionChange)change).version().isAfter(deployment.version())) return false; // later is ok
```

Comment:
I'd prefer if the order of the versions (change vs deployment) was the same as the order of the revisions (deployment vs change).

Reply:
Also, the doc should state that the jobs are checked for success on the currently deploying change, or on a later version, if currently upgrading.

Reply:
I guess the success (of the deployment job) isn't even a factor anymore? If a new version/revision has been deployed, and tests have then failed, it will still count as a success here.

Reply:
Both done

Reply:
:)